jujugui/static/gui/src/app/jujulib/test-bakery.js

```diff
@@ -250,5 +261,133 @@ describe('Bakery', () => {
       assert.equal(args[0], 'macaroon');
     });
   });
+
+  describe('interact handling', () => {
+    const getResponse = fail => {
+      let response = {
+        target: {
+          status: 200,
+          responseText: '',
+          response: '',
+        }
+      };
+      if (fail) {
+        response.target.status = 0;
+      }
+      return response;
+    };
+
+    it('accepts a visit page method', () => {
+      const params = {
+        visitPage: () => { return 'visits'; }
+      };
+      bakery = new jujulib.Bakery(client, storage, params);
+      assert.equal(bakery._visitPage(), 'visits');
+    });
+
+    it('opens the visit page', () => {
+      const windowStub = sinon.stub(window, 'open');
```

Comment:
This is a good example of why I like to pass things like 'window' and 'document' into the constructors. This will stub the `window.open` method. But if the test exits before the restore. it'll remain stubbed and you'll could have cascading failures elsewhere in the test run and not know why.